presto-main/src/main/java/com/facebook/presto/operator/scalar/MapGenericEquality.java

```diff
@@ -41,29 +40,23 @@ public static Boolean genericEqual(
             Block rightMapBlock,
             EqualityPredicate predicate)
     {
-        Map<KeyWrapper, Integer> wrappedLeftMap = new LinkedHashMap<>();
-        for (int position = 0; position < leftMapBlock.getPositionCount(); position += 2) {
-            wrappedLeftMap.put(new KeyWrapper(readNativeValue(keyType, leftMapBlock, position), keyEqualsFunction, keyHashcodeFunction), position + 1);
-        }
-
-        Map<KeyWrapper, Integer> wrappedRightMap = new LinkedHashMap<>();
-        for (int position = 0; position < rightMapBlock.getPositionCount(); position += 2) {
-            wrappedRightMap.put(new KeyWrapper(readNativeValue(keyType, rightMapBlock, position), keyEqualsFunction, keyHashcodeFunction), position + 1);
-        }
-
-        if (wrappedLeftMap.size() != wrappedRightMap.size()) {
+        if (leftMapBlock.getPositionCount() != rightMapBlock.getPositionCount()) {
             return false;
         }
 
-        for (Map.Entry<KeyWrapper, Integer> entry : wrappedRightMap.entrySet()) {
-            KeyWrapper key = entry.getKey();
-            Integer leftValuePosition = wrappedLeftMap.get(key);
-            if (leftValuePosition == null) {
+        SingleMapBlock mapLeftBlock = (SingleMapBlock) leftMapBlock;
+        SingleMapBlock mapRightBlock = (SingleMapBlock) rightMapBlock;
+
+        for (int position = 0; position < leftMapBlock.getPositionCount(); position += 2) {
+            Object key = readNativeValue(keyType, mapLeftBlock, position);
+            int leftPosition = mapLeftBlock.seekKey(key);
```

Comment:
`int leftPosition = position + 1`

Map block in Presto is laid out as `k1, v1, k2, v2, k3, v3, ...`. This is part of the contract of the map block. You can safely assume that. Therefore, it is unnecessary to call seekKey on the blocks.